File: Algorithms/manachers.cpp
string manachers_algorithm(string &s) {
    if (s.empty()) {
        return "";
    }
    string t = "^#";

    for (auto c : s) {
        t.push_back(c);
        t.push_back('#');
    }
    t.push_back('$');
        
    vector<int> radii(t.size());
    int c = 1, d = 1;
    
    for(int i = 2; i < radii.size()-1; ++i) {
        int mirror = c-(i-c);
        radii[i] = max(0, min(radii[mirror], d-i));
        while(t[i-radii[i]-1] == t[i+radii[i]+1]) {
            radii[i]++;
        }
        if (radii[i] + i > d) {
            c = i;
            d = i + radii[i];
        }
    }
    
    int center = 1, max_rad = radii[1];
    for (int i = 1; i < radii.size()-1; ++i) {
        if (radii[i] > max_rad) {
            center = i;
            max_rad = radii[i];
        }
    }
    
    int start = (center - max_rad) / 2;
    int end = (center + max_rad) / 2;
    
    return s.substr(start, end-start);
    
}

File: Algorithms/manachers.py
def manachers_algorithm(s):
    if not s:
        return ""
    t = "^#" + "#".join(s) + "#$"
    c = 1
    d = 1
    p = [0] * len(t)
    for i in range(2, len(t)-1):
        mirror =  c-(i-c)
        p[i] = max(0, min(p[mirror], d-i))
        while t[i-1-p[i]] == t[i+p[i]+1]:
            p[i] += 1
            
        if i + p[i] > d:
            c = i
            d = i + p[i]
            
    (k, i) = max((p[i], i) for i in range(1, len(t)-1))
    start = (i-k)//2
    end = (i+k)//2
    return s[start:end]